File: rust/examples/hello_world.rs
//! Simple example that loads and executes a "Hello World" LoongArch program

use libloong::{Machine, MachineOptions};
use std::env;
use std::fs;
use std::process;
use std::time::Instant;

fn main() {
    // Get ELF path from command line
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <loongarch_elf_file>", args[0]);
        eprintln!();
        eprintln!("Example:");
        eprintln!("  {} ../tests/programs/hello_world.elf", args[0]);
        process::exit(1);
    }

    let elf_path = &args[1];

    // Read the ELF binary
    let binary = match fs::read(elf_path) {
        Ok(data) => data,
        Err(e) => {
            eprintln!("Failed to read ELF file '{}': {}", elf_path, e);
            process::exit(1);
        }
    };

    println!("Loading LoongArch ELF: {}", elf_path);
    println!("Binary size: {} bytes", binary.len());

    // Create machine with custom options
    let options = MachineOptions {
        memory_max: 600 * 1024 * 1024,
        ..Default::default()
    };
    let mut machine = match Machine::new(&binary, options) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("Failed to create machine: {}", e);
            process::exit(1);
        }
    };

    // Setup stdout callback to capture output
    libloong::set_stdout_callback(Some(|data| {
        print!("{}", String::from_utf8_lossy(data));
    }));

    // Setup Linux environment
    Machine::setup_linux_syscalls();
    if let Err(e) = machine.setup_linux(&["program"], &[]) {
        eprintln!("Failed to setup Linux: {}", e);
        process::exit(1);
    }

    // Execute the program (unlimited instructions)
    let start_time = Instant::now();
    let max_instructions = u64::MAX;

    match machine.simulate(max_instructions) {
        Ok(_) => {
            let elapsed = start_time.elapsed();
            let exit_code = machine.return_value() as i32; // $a0 (r4) contains exit code

            println!(
                "Exit code: {}  Time: {:.3} seconds",
                exit_code,
                elapsed.as_secs_f64()
            );

            // Only show instruction count if we used a limited instruction count
            // (unlimited uses faster inaccurate dispatch with no counting)
            if max_instructions != u64::MAX {
                let instr_count = machine.instruction_counter();
                let mips = instr_count as f64 / (elapsed.as_secs_f64() * 1_000_000.0);
                println!("Instructions: {} ({:.2} MI/s)", instr_count, mips);
            }

            process::exit(exit_code);
        }
        Err(e) => {
            let elapsed = start_time.elapsed();
            eprintln!("\nExecution failed: {}", e);
            eprintln!("Time: {:.3} seconds", elapsed.as_secs_f64());

            if max_instructions != u64::MAX {
                eprintln!("Instructions executed: {}", machine.instruction_counter());
            }
            eprintln!("PC: 0x{:016x}", machine.get_pc());
            process::exit(1);
        }
    }
}

File: go/examples/hello_world.go
// Example that loads and executes a "Hello World" LoongArch program
package main

import (
	"fmt"
	"os"
	"time"

	libloong "github.com/gonz/libloong/go"
)

func main() {
	// Get ELF path from command line
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <loongarch_elf_file>\n", os.Args[0])
		fmt.Fprintln(os.Stderr)
		fmt.Fprintln(os.Stderr, "Example:")
		fmt.Fprintf(os.Stderr, "  %s ../tests/programs/hello_world.elf\n", os.Args[0])
		os.Exit(1)
	}

	elfPath := os.Args[1]

	// Read the ELF binary
	binary, err := os.ReadFile(elfPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read ELF file '%s': %v\n", elfPath, err)
		os.Exit(1)
	}

	fmt.Printf("Loading LoongArch ELF: %s\n", elfPath)
	fmt.Printf("Binary size: %d bytes\n", len(binary))

	// Create machine with custom options
	options := libloong.DefaultOptions()
	options.MemoryMax = 600 * 1024 * 1024

	machine, err := libloong.NewMachine(binary, &options)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create machine: %v\n", err)
		os.Exit(1)
	}
	defer machine.Close()

	// Setup Linux environment
	libloong.SetupLinuxSyscalls()
	if err := machine.SetupLinux([]string{"program"}, []string{}); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to setup Linux: %v\n", err)
		os.Exit(1)
	}

	// Execute the program (unlimited instructions)
	startTime := time.Now()
	const maxInstructions = ^uint64(0) // unlimited

	err = machine.Simulate(maxInstructions)
	elapsed := time.Since(startTime)

	if err != nil {
		fmt.Fprintf(os.Stderr, "\nExecution failed: %v\n", err)
		fmt.Fprintf(os.Stderr, "Time: %.3f seconds\n", elapsed.Seconds())

		if maxInstructions != ^uint64(0) {
			fmt.Fprintf(os.Stderr, "Instructions executed: %d\n", machine.InstructionCounter())
		}
		fmt.Fprintf(os.Stderr, "PC: 0x%016x\n", machine.GetPC())
		os.Exit(1)
	}

	exitCode := int(machine.ReturnValue()) // $a0 contains exit code

	fmt.Printf("Exit code: %d  Time: %.3f seconds\n", exitCode, elapsed.Seconds())

	// Only show instruction count if we used a limited instruction count
	// (unlimited uses faster inaccurate dispatch with no counting)
	if maxInstructions != ^uint64(0) {
		instrCount := machine.InstructionCounter()
		mips := float64(instrCount) / (elapsed.Seconds() * 1_000_000.0)
		fmt.Printf("Instructions: %d (%.2f MI/s)\n", instrCount, mips)
	}

	os.Exit(exitCode)
}
